Sources/OpenCombineJS/JSPromise.swift

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 OpenCombineJS contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import JavaScriptKit
+import OpenCombine
+
+extension JSPromise where Success: JSValueConstructible, Failure: JSError {
+  public final class PromisePublisher: Publisher {
+    public typealias Output = Success
+
+    /// Reference to a parent promise instance to prevent early deallocation
+    private var parent: JSPromise?
+
+    /// Reference to a `then` success callback promise instance to prevent early deallocation
+    private var then: JSPromise<JSValue, Failure>?
+
+    /// `Future` instance that handles subscriptions to this publisher.
+    private var future: Future<Success, Failure>?
+
+    fileprivate init(parent: JSPromise) {
+      future = .init { resolver in
```

Comment:
```suggestion
      future = .init { [weak self] resolver in
```

Need to capture `self` weakly here

Reply:
I've tried that, but it didn't make any difference, `Future` doesn't capture this closure at all and executes it immediately. It shouldn't be marked `@escaping` in the first place, but was marked so for compatibility with Apple's Combine I assume.

Reply:
That depends on their implementation and capture behavior can be changed. So I think it's better to capture self weakly.

Reply:
Makes sense, that's updated now 🙂